File: python/src/gibberish_rewriter/core/key_names.py
"""Key names and virtual-key codes, loaded from shared/keys.json."""

from __future__ import annotations

import enum
import json
from pathlib import Path


class HoldKeys(enum.IntFlag):
    """The keys a chord can hold. Flags, so a chord's hold set is one value."""

    NONE = 0
    SHIFT = 1
    CTRL = 2
    ALT = 4
    WIN = 8
    CAPS_LOCK = 16


VK_BACKSPACE = 0x08
VK_CAPS_LOCK = 0x14
VK_PACKET = 0xE7
VK_MASK = 0xE8
"""Unassigned key pressed after an Alt or Win chord fires."""

HOLD_KEY_ORDER = (HoldKeys.SHIFT, HoldKeys.CTRL, HoldKeys.ALT, HoldKeys.WIN, HoldKeys.CAPS_LOCK)
"""The order hold keys are written in (canonical hotkey text)."""

HOLD_KEY_NAMES = {
    HoldKeys.SHIFT: "Shift",
    HoldKeys.CTRL: "Ctrl",
    HoldKeys.ALT: "Alt",
    HoldKeys.WIN: "Win",
    HoldKeys.CAPS_LOCK: "CapsLock",
}
"""Hold-key names as keys.json and hotkey strings write them."""

_ASCII_LOWER = str.maketrans("ABCDEFGHIJKLMNOPQRSTUVWXYZ", "abcdefghijklmnopqrstuvwxyz")


def ascii_lower(text: str) -> str:
    """Lowercases A-Z only."""
    return text.translate(_ASCII_LOWER)


class KeyNames:
    """Key names and virtual-key codes, loaded from shared/keys.json."""

    def __init__(
        self,
        triggers: dict[str, int],
        modifiers: dict[str, int],
        hold_vks: dict[HoldKeys, list[int]],
        table_keys: list[str],
    ) -> None:
        self._triggers = {ascii_lower(name): vk for name, vk in triggers.items()}
        self._modifiers = {ascii_lower(name): vk for name, vk in modifiers.items()}
        self._hold_vks = hold_vks
        self._hold_by_name = {ascii_lower(HOLD_KEY_NAMES[hold]): hold for hold in hold_vks}
        self._hold_by_vk: dict[int, HoldKeys] = {}
        self._modifier_vks: set[int] = set()
        self._name_by_vk: dict[int, str] = {}
        for hold, vks in hold_vks.items():
            for vk in vks:
                self._hold_by_vk[vk] = hold
                self._modifier_vks.add(vk)
        for name, vk in modifiers.items():
            self._modifier_vks.add(vk)
            self._name_by_vk.setdefault(vk, name)
        for name, vk in triggers.items():
            self._name_by_vk.setdefault(vk, name)
        self.table_keys: tuple[str, ...] = tuple(table_keys)
        """The key-table keys in table order, which is also the snapshot's order. LayoutReader reads these."""

    @classmethod
    def load(cls, path: str | Path) -> KeyNames:
        return cls.parse(Path(path).read_text(encoding="utf-8"))

    @classmethod
    def parse(cls, text: str) -> KeyNames:
        root = json.loads(text)
        names_to_hold = {name: hold for hold, name in HOLD_KEY_NAMES.items()}
        hold_vks = {names_to_hold[name]: list(codes) for name, codes in root["holdKeys"].items()}
        triggers = dict(root["keys"])
        table_keys = list(root["tableKeys"])
        lowered = {ascii_lower(name) for name in triggers}
        unknown = next((name for name in table_keys if ascii_lower(name) not in lowered), None)
        if unknown is not None:
            raise ValueError(f'tableKeys lists "{unknown}", which is not in keys.')
        return cls(triggers, dict(root["modifiers"]), hold_vks, table_keys)

    def trigger_vk(self, name: str) -> int | None:
        """Looks up a non-modifier key name, case-insensitively. "`" means Backquote."""
        return self._triggers.get(ascii_lower("Backquote" if name == "`" else name))

    def vk_of(self, name: str) -> int | None:
        """Looks up any key name, modifier or not."""
        vk = self.trigger_vk(name)
        return vk if vk is not None else self._modifiers.get(ascii_lower(name))

    def hold_key(self, name: str) -> HoldKeys | None:
        return self._hold_by_name.get(ascii_lower(name))

    def is_modifier_name(self, name: str) -> bool:
        """True for hold-key names such as Shift and modifier key names such as ShiftLeft or NumLock."""
        lowered = ascii_lower(name)
        return lowered in self._hold_by_name or lowered in self._modifiers

    def is_modifier(self, vk: int) -> bool:
        return vk in self._modifier_vks

    def hold_key_of(self, vk: int) -> HoldKeys:
        return self._hold_by_vk.get(vk, HoldKeys.NONE)

    def vks_of(self, holds: HoldKeys) -> list[int]:
        """Codes of every hold key in holds, in canonical order."""
        return [vk for hold in HOLD_KEY_ORDER if hold in holds for vk in self._hold_vks[hold]]

    def name_of(self, vk: int) -> str:
        return self._name_by_vk.get(vk, f"0x{vk:02X}")

File: csharp/src/GibberishRewriter.Core/KeyNames.cs
using System.Text.Json;

namespace GibberishRewriter.Core;

/// <summary>The keys a chord can hold. Flags, so a chord's hold set is one value.</summary>
[Flags]
public enum HoldKeys
{
    None = 0,
    Shift = 1,
    Ctrl = 2,
    Alt = 4,
    Win = 8,
    CapsLock = 16,
}

/// <summary>Key names and virtual-key codes, loaded from shared/keys.json.</summary>
public sealed class KeyNames
{
    public const int VkBackspace = 0x08;
    public const int VkCapsLock = 0x14;
    public const int VkPacket = 0xE7;

    /// <summary>Unassigned key pressed after an Alt or Win chord fires.</summary>
    public const int VkMask = 0xE8;

    /// <summary>The order hold keys are written in (canonical hotkey text).</summary>
    public static IReadOnlyList<HoldKeys> HoldKeyOrder { get; } =
        [HoldKeys.Shift, HoldKeys.Ctrl, HoldKeys.Alt, HoldKeys.Win, HoldKeys.CapsLock];

    private readonly Dictionary<string, int> _triggers;
    private readonly Dictionary<string, int> _modifiers;
    private readonly Dictionary<HoldKeys, int[]> _holdVks;
    private readonly Dictionary<string, HoldKeys> _holdByName = new(StringComparer.OrdinalIgnoreCase);
    private readonly Dictionary<int, HoldKeys> _holdByVk = [];
    private readonly HashSet<int> _modifierVks = [];
    private readonly Dictionary<int, string> _nameByVk = [];

    private KeyNames(
        Dictionary<string, int> triggers,
        Dictionary<string, int> modifiers,
        Dictionary<HoldKeys, int[]> holdVks,
        string[] tableKeys)
    {
        _triggers = triggers;
        TableKeys = tableKeys;
        _modifiers = modifiers;
        _holdVks = holdVks;
        foreach (var (hold, vks) in holdVks)
        {
            _holdByName[hold.ToString()] = hold;
            foreach (var vk in vks)
            {
                _holdByVk[vk] = hold;
                _modifierVks.Add(vk);
            }
        }
        foreach (var (name, vk) in modifiers)
        {
            _modifierVks.Add(vk);
            _nameByVk.TryAdd(vk, name);
        }
        foreach (var (name, vk) in triggers)
        {
            _nameByVk.TryAdd(vk, name);
        }
    }

    /// <summary>The key-table keys in table order, which is also the snapshot's order. LayoutReader reads these.</summary>
    public IReadOnlyList<string> TableKeys { get; }

    public static KeyNames Load(string path) => Parse(File.ReadAllText(path));

    public static KeyNames Parse(string json)
    {
        using var document = JsonDocument.Parse(json);
        var root = document.RootElement;
        var holdVks = new Dictionary<HoldKeys, int[]>();
        foreach (var property in root.GetProperty("holdKeys").EnumerateObject())
        {
            holdVks[Enum.Parse<HoldKeys>(property.Name)] =
                property.Value.EnumerateArray().Select(code => code.GetInt32()).ToArray();
        }
        var triggers = ReadCodes(root.GetProperty("keys"));
        var tableKeys = root.GetProperty("tableKeys").EnumerateArray().Select(name => name.GetString()!).ToArray();
        if (tableKeys.FirstOrDefault(name => !triggers.ContainsKey(name)) is { } unknown)
        {
            throw new FormatException($"tableKeys lists \"{unknown}\", which is not in keys.");
        }
        return new KeyNames(triggers, ReadCodes(root.GetProperty("modifiers")), holdVks, tableKeys);
    }

    private static Dictionary<string, int> ReadCodes(JsonElement element)
    {
        var codes = new Dictionary<string, int>(StringComparer.OrdinalIgnoreCase);
        foreach (var property in element.EnumerateObject())
        {
            codes[property.Name] = property.Value.GetInt32();
        }
        return codes;
    }

    /// <summary>Looks up a non-modifier key name, case-insensitively. "`" means Backquote.</summary>
    public bool TryGetTrigger(string name, out int vk) =>
        _triggers.TryGetValue(name == "`" ? "Backquote" : name, out vk);

    /// <summary>Looks up any key name, modifier or not.</summary>
    public bool TryGetVk(string name, out int vk) =>
        TryGetTrigger(name, out vk) || _modifiers.TryGetValue(name, out vk);

    public bool TryGetHoldKey(string name, out HoldKeys hold) => _holdByName.TryGetValue(name, out hold);

    /// <summary>True for hold-key names such as Shift and modifier key names such as ShiftLeft or NumLock.</summary>
    public bool IsModifierName(string name) => _holdByName.ContainsKey(name) || _modifiers.ContainsKey(name);

    public bool IsModifier(int vk) => _modifierVks.Contains(vk);

    public HoldKeys HoldKeyOf(int vk) => _holdByVk.TryGetValue(vk, out var hold) ? hold : HoldKeys.None;

    /// <summary>Codes of every hold key in <paramref name="holds"/>, in canonical order.</summary>
    public IEnumerable<int> VksOf(HoldKeys holds) =>
        HoldKeyOrder.Where(hold => holds.HasFlag(hold)).SelectMany(hold => _holdVks[hold]);

    public string NameOf(int vk) => _nameByVk.TryGetValue(vk, out var name) ? name : $"0x{vk:X2}";
}
